File: package_generator_templates/templates/cpp_service_server/template/ros/src/node_ros.cpp
/**
* @file {nodeName}_ros.cpp
* @author {packageAuthor}
*
* Copyright (C) {packageCopyright}
*
* @brief {packageDescription}
* @warning This file should not be edited
**/

// ROS includes
#include <ros/ros.h>

{ifdynParameter}
#include <dynamic_reconfigure/server.h>
#include <{packageName}/{nodeName}Config.h>

{endifdynParameter}
// ROS message & services includes
{foralldirectPublisher}
#include <{apply-get_cpp_path}.h>
{endforalldirectPublisher}
{foralldirectSubscriber}
#include <{apply-get_cpp_path}.h>
{endforalldirectSubscriber}
{forallserviceServer}
#include <{apply-get_cpp_path}.h>
{endforallserviceServer}
{forallserviceClient}
#include <{apply-get_cpp_path}.h>
{endforallserviceClient}

// other includes
#include <{nodeName}_common.cpp>

/**
 * @class {apply-capitalized_node_name}ROS
 * @brief Class handling the connection with the ROS world.
 * It also implement the node life-cycle, and access to object {nodeName}-impl when appropriate
 */
class {apply-capitalized_node_name}ROS
{
public:
    ros::NodeHandle n_;
    ros::NodeHandle np_;

    {ifdynParameter}
    dynamic_reconfigure::Server<{packageName}::{nodeName}Config> server;
    dynamic_reconfigure::Server<{packageName}::{nodeName}Config>::CallbackType f;
    {endifdynParameter}
    {forallserviceServer}
    ros::ServiceServer {name}_;
    {endforallserviceServer}
    // todo confirm it should be always defined, even if not used.
    {apply-capitalized_node_name}Config component_config_;
    {apply-capitalized_node_name}Impl component_implementation_;

    /**
     * @brief object constructor.
     */
    {apply-capitalized_node_name}ROS() : np_("~")
    {
        {ifdynParameter}
        // preparing dynamic reconfigure mechanism
        f = boost::bind(&{apply-capitalized_node_name}ROS::configure_callback, this, _1, _2);
        server.setCallback(f);
        {endifdynParameter}
        {ifdirectPublisher}
        // Handling direct publisher
        {foralldirectPublisher}
        component_implementation_.passthrough.{name} = n_.advertise<{type}>("{name}", 1);
        {endforalldirectPublisher}
        {endifdirectPublisher}
        {ifdirectSubscriber}
        // Handling direct subscriber
        {foralldirectSubscriber}
        component_implementation_.passthrough.{name} = n_.subscribe("{name}", 1, &{apply-capitalized_node_name}Impl::directTopicCallback_{name}, &component_implementation_);
        {endforalldirectSubscriber}
        {endifdirectSubscriber}
        {ifparameter}
        // handling parameters from the parameter server
        {endifparameter}
        {forallparameter}
        np_.param("{name}", component_config_.{name}, ({type}){apply-get_cpp_param_value});
        {endforallparameter}
        {ifdynParameter}
        // handling dynamic parameters
        {endifdynParameter}
        {foralldynParameter}
        np_.param("{name}", component_config_.{name}, ({type}){apply-get_cpp_param_value});
        {endforalldynParameter}
        {ifserviceServer}
        // handling Service servers
        {endifserviceServer}
        {forallserviceServer}
        {name}_ = n_.advertiseService<{type}::Request , {type}::Response>("{name}", boost::bind(&{apply-capitalized_node_name}Impl::callback_{name}, &component_implementation_,_1,_2,component_config_));
        {endforallserviceServer}
        {ifserviceClient}
        // handling Service clients
        {endifserviceClient}
        {forallserviceClient}
        component_implementation_.passthrough.client_{name} = n_.serviceClient<{type}>("{name}");
        {endforallserviceClient}
    }

    {ifdynParameter}
    /**
     * @brief callback called when a dynamically reconfigurable parameter is changed
     * @param config the parameter structure to be updated
     * @param level level of parameters changed
     */
    void configure_callback({packageName}::{nodeName}Config &config, uint32_t level)
    {
        {foralldynParameter}
        component_config_.{name} = config.{name};
        {endforalldynParameter}
    }

    {endifdynParameter}
    /**
     * @brief configure function called after node creation.
     */
    bool configure()
    {
        return component_implementation_.configure(component_config_);
    }
    /**
     * @brief object destructor
     */
    ~{apply-capitalized_node_name}ROS()
    {
    }
};

/**
 * @brief Main of the component. Create the object and launches the periodical call to upde method.
 */
int main(int argc, char** argv)
{

    ros::init(argc, argv, "{nodeName}");

    {apply-capitalized_node_name}ROS node;
    if (!node.configure())
    {
        ROS_FATAL("Could not configure the node");
        ROS_FATAL("Please check configuration parameters");
        return -1;
    }

    ros::spin();

    return 0;
}

File: package_generator_templates/templates/python_service_server/template/src/package_name/node_ros.py
#!/usr/bin/env python
"""
@package {packageName}
@file {nodeName}_ros.py
@author {packageAuthor}
@brief {packageDescription}

Copyright (C) {packageCopyright}
{packageLicense}
"""

from copy import deepcopy
import rospy
{ifdynParameter}
from dynamic_reconfigure.server import Server
from {packageName}.cfg import {nodeName}Config
{endifdynParameter}

# ROS message & services includes
{forallserviceServer}
from {apply-get_package_type}.srv import {apply-get_class_type}
{endforallserviceServer}
# other includes
from {packageName} import {nodeName}_impl


class {apply-capitalized_node_name}ROS(object):
    """
    ROS interface class, handling all communication with ROS
    """
    def __init__(self):
        """
        Attributes definition
        """
        self.component_config_ = {nodeName}_impl.{apply-capitalized_node_name}Config()
        self.component_implementation_ = {nodeName}_impl.{apply-capitalized_node_name}Implementation()

        {ifdynParameter}
        # preparing dynamic reconfigure mechanism
        srv = Server({nodeName}Config, self.configure_callback)
        {endifdynParameter}
        {ifparameter}
        # handling parameters from the parameter server
        {endifparameter}
        {forallparameter}
        self.component_config_.{name} = rospy.get_param("~{name}", {apply-get_py_param_value})
        {endforallparameter}
        {ifdynParameter}
        # handling dynamic parameters
        {endifdynParameter}
        {foralldynParameter}
        self.component_config_.{name} = rospy.get_param("{name}", {apply-get_py_param_value})
        {endforalldynParameter}
        {ifserviceServer}
        # handling service servers
        {endifserviceServer}
        {forallserviceServer}
        self.{name}_ = rospy.Service('{name}', {apply-get_class_type}, self.component_implementation_.callback_{name})
        {endforallserviceServer}

    {ifdynParameter}
    def configure_callback(self, config, level):
        """
        callback on the change of parameters dynamically adjustable
        """
        {foralldynParameter}
        self.component_config_.{name} = config.{name}
        {endforalldynParameter}
        return config

    {endifdynParameter}
    def configure(self):
        """
        function setting the initial configuration of the node
        """
        return self.component_implementation_.configure(self.component_config_)


def main():
    """
    @brief Entry point of the package.
    Instanciate the node interface containing the Developer implementation
    @return nothing
    """
    rospy.init_node("{nodeName}", anonymous=False)

    node = {apply-capitalized_node_name}ROS()
    if not node.configure():
        rospy.logfatal("Could not configure the node")
        rospy.logfatal("Please check configuration parameters")
        rospy.logfatal("{}".format(node.component_config_))
        return

    rospy.spin()
